### DISCUSSION TITLE
How to build with LibCarla and UE4 content with ROS2

### DISCUSSION THREAD

**Discussion by pravinkmr26** [carla-simulator/carla#7254] [Q&A]
Hi team, I see WITH_ROS2 c++ macro is used in many places, especially the sensor outputs. when I tried enabling the ros2 flag in linux make file I get a bunch of compiler errors, llike fastdds headers are missing, etc. while I can solve them manually, I want to understand, is there any script or specific steps that I need to follow to enable ROS2?

**Comment by xavisolesoft** [comment]
You just need to build and launch CARLA with the next command:
```
make launch ARGS="--ros2"
```

Note that this will enable the macro `WITH_ROS2`, so you do not need to enable it by hand.

**Reply by pravinkmr26** [reply]
Yes, thanks. I figured it out. but when I was looking at the code I see the Ros publishers use dds types instead of ROS types. I would like to understand how the subscriber would use this data. I was hoping that the data would be published to ROS types (say sensor_msgs::PointCloud2 for ex). can you direct me to some design docs so that I can understand how publishing over ROS would work from a high level. (publisher&subscriber flow)

**Reply by pravinkmr26** [reply]
or do you think this is a different discussion?

**Reply by xavisolesoft** [reply]
I think is a good idea to open a new discussion for this new topic.
This will help other people to find the discussion.

**Comment by uff1199** [comment]
For others: In CARLA 0.9.16 ros2 middleware is supported. During sensor spawning via the python API only enable_for_ros() needs to be called for each actor. The topics can be then subscribed as ROS2 topics, which I tested with ROS2 Humble.
